artemis-commons/src/main/java/org/apache/activemq/artemis/utils/collections/LinkedListImpl.java

```diff
@@ -46,13 +49,59 @@
 
    private final Comparator<E> comparator;
 
+   private ToLongFunction<E> idSupplier;
+
+   LongObjectHashMap<Node<E>> nodeMap;
+
    public LinkedListImpl() {
-      this(null);
+      this(null, null);
    }
 
+
    public LinkedListImpl(Comparator<E> comparator) {
+      this(comparator, null);
+   }
+
+   @Override
+   public void clearID() {
+      idSupplier = null;
+      nodeMap.clear(); // just a little hand to GC
+      nodeMap = null;
+   }
+
+   @Override
+   public void setIDSupplier(ToLongFunction<E> supplier) {
+      this.idSupplier = supplier;
+      nodeMap = new LongObjectHashMap<>(this.size);
```

Comment:
What happen to any previously set `idSupplier`/`nodeMap`?
Better to `throw new IllegalStateException("cannot set an ID supplier twice")` to save it to happen

Reply:
If I did that, I would just change the implementation of queue to call clear before setting the new ID. 

I would prefer to keep it that way.